unittests/runtime-collections/rangemaps.cpp

```diff
@@ -0,0 +1,596 @@
+#include <boost/test/unit_test.hpp>
+
+#include <runtime/collections/rangemap.h>
+
+BOOST_AUTO_TEST_SUITE(RangeMapTest)
+
+BOOST_AUTO_TEST_CASE(rangemap_test_empty) {
+  auto map = RangeMap<int, int>();
+  auto result = map.size();
+  BOOST_CHECK_EQUAL(result, 0);
+  auto result2 = map.containsKey(0);
+  BOOST_CHECK_EQUAL(result2, false);
+}
+
+BOOST_AUTO_TEST_CASE(rangemap_test_concat_success) {
+  auto m = (RangeMap<int, int>()).inserted(std::make_pair(2, 4), 1);
+  auto map1 = m.inserted(std::make_pair(6, 7), 1);
+  auto m1 = (RangeMap<int, int>()).inserted(std::make_pair(0, 2), 2);
+  auto m2 = m1.inserted(std::make_pair(4, 6), 3);
+  auto map2 = m2.inserted(std::make_pair(8, 9), 4);
+  auto map3 = concat(map1, map2);
+  auto result = map3.size();
+  BOOST_CHECK_EQUAL(result, 5);
+  auto map4 = (RangeMap<int, int>()).inserted(std::make_pair(4, 6), 1);
+  auto map5 = concat(map1, map4);
+  result = map5.size();
+  BOOST_CHECK_EQUAL(result, 1);
+}
+
+BOOST_AUTO_TEST_CASE(rangemap_test_concat_failure) {
+  auto m = (RangeMap<int, int>()).inserted(std::make_pair(2, 4), 1);
+  auto map1 = m.inserted(std::make_pair(6, 7), 2);
+  auto m1 = (RangeMap<int, int>()).inserted(std::make_pair(0, 3), 2);
+  auto map2 = m1.inserted(std::make_pair(8, 9), 4);
+  BOOST_CHECK_THROW(concat(map1, map2), std::invalid_argument);
```

Comment:
now I'm particularly confused. You appear to be concatenating two maps with overlapping ranges and it's generating an error, but the same thing actually happens in the previous test for the range `[4, 6)` and no error occurs. Again, am I missing something?

Reply:
Right, concatenating two maps with overlapping ranges generates an error. I followed the unordered map implementation in K for this, where if the maps to concatenate share a key an error is generated. There is no contradiction with the previous example, as there were no overlapping ranges in that case.